File: day38/C++/stack.cpp
#include<iostream>
#include<limits>
#include<array>
#include "stack.h"
#define SIZE 6

template<class T, const int MAX_SIZE>
bool Stack<T, MAX_SIZE>::isEmpty() {
  return this -> size == 0;
}

template<class T, const int MAX_SIZE>
bool Stack<T, MAX_SIZE>::isFull() {
  return this -> size == MAX_SIZE;
}

template<class T, const int MAX_SIZE>
void Stack<T, MAX_SIZE>::push(T element) {
  if (this -> isFull()) {
    std::cerr << "Cannot push an element into a full stack\n";
    return;
  }
  data[this -> size - 1] = element;
  ++this->size;
}

template<class T, const int MAX_SIZE>
T Stack<T, MAX_SIZE>::pop() {
  if (this -> isEmpty()) {
    std::cerr << "Cannot pop an element from an empty stack\n";
    return NULL;
  }
  T element = this -> peek();
  data[this -> size - 1] = NULL;
  --this -> size;
  return element;
}

template<class T, const int MAX_SIZE>
T Stack<T, MAX_SIZE>::peek() {
  if (this -> isEmpty()) {
    std::cerr << "No element in the stack\n";
    return NULL;
  }
  T element = data[this -> size - 2];
  return element;
}

int main() {
  Stack<int, SIZE> int_stack;
  std::cout << (int_stack.isEmpty() ? "Stack is empty" : "Stack is full") << "\n";
  int element = int_stack.pop();
  element = int_stack.peek();
  std::array<int, SIZE> a{3, 4, 9, 8, 5, 6};
  for (int i = 0; i < SIZE; i++) {
    int_stack.push(a[i]);
  }
  std::cout << (int_stack.isFull() ? "Stack is full" : "Stack is empty") << "\n";
  int_stack.push(9);
  std::cout << int_stack.pop() << "\n";
  std::cout << int_stack.peek() << "\n";
  return 0;
}

File: day38/Java/Stack.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package javaapplication3;

/**
 * @date 11/02/19
 * @author SPREEHA DUTTA
 */
import java.util.*;
public class Stack {
    static int st[]=new int[5];
    static int top=0;
    public static void push(int n)
    {
        if(top<5)
        {
            st[top]=n;
            top++;
        }
        else
            isFull();
    }
    public static void pop()
    {
        if(top<=5 && top>0)
        {
            top--;
            System.out.println("Popped element is "+st[top]);
        }
        else
            isEmpty();
    }
    public static void isEmpty()
    {
        if(top<=0)
            System.out.println("Stack is empty");
    }
    public static void isFull()
    {
        if(top>=5)
            System.out.println("Stack is full");
    }
    public static void peek()
    {
        if(top-1>=0 && top-1<5)
            System.out.println("The topmost element of the stack is "+st[top-1]);
        else
          System.out.println("Stack is empty");  
    }
    public static void display()
    {
        System.out.println("All the elements present in the stack are:");
        for(int i=0;i<top;i++)
            System.out.print(st[i]+" ");
        System.out.println();
    }
    public static void main(String []args)
    {
        Scanner sc=new Scanner(System.in);
        int ch;int c=0;
        do
        {
            c=0;
            System.out.println("Enter choice ");
            System.out.println("1. Add new element"
                    + "\n"+"2. Remove topmost eleemnt"
            +"\n3. To view the topmost element"+"\n4. To view all the elements");
            ch=sc.nextInt();
            switch(ch)
            {
                case 1:
                    System.out.println("Enter element to be added to stack ");
                    int n=sc.nextInt();
                    push(n);
                    break;
                case 2:
                    pop();
                    break;
                case 3:
                    peek();
                    break;
                case 4:
                    display();
                    break;
                default:
                    System.out.println("Invalid choice :");
            }
            System.out.println("Enter 1 if you wish to continue");
            c=sc.nextInt();
        }
        while(c==1);
    }
}
